netty/src/main/java/io/grpc/netty/ConfigurableX509TrustManager.java

```diff
@@ -0,0 +1,148 @@
+      throws CertificateException {
+
+  }
+
+  @Override
+  public void checkServerTrusted(X509Certificate[] x509Certificates, String s, SSLEngine sslEngine)
+      throws CertificateException {
+    checkTrusted(x509Certificates, s, sslEngine, true);
+  }
+
+  @Override
+  public void checkServerTrusted(X509Certificate[] x509Certificates, String s)
+      throws CertificateException {
+
+  }
+
+  @Override
+  public X509Certificate[] getAcceptedIssuers() {
+    return new X509Certificate[0];
+  }
+
+  private void checkTrusted(X509Certificate[] x509Certificates, String s, SSLEngine sslEngine,
+      boolean isClient) throws CertificateException {
+    VerificationAuthType authType = this.tlsOptions.getVerificationAuthType();
+    if (authType == VerificationAuthType.CertificateAndHostNameVerification
+        || authType == VerificationAuthType.CertificateVerification) {
+      if (x509Certificates == null || x509Certificates.length == 0) {
+        throw new CertificateException(
+            "Want certificate verification but got null or empty certificates");
+      }
+      KeyStore ks;
+      try {
+        ks = this.tlsOptions.getTrustedCerts();
+      } catch (Exception e) {
+        throw new CertificateException("Function getTrustedCerts fails, error: " + e.getMessage());
+      }
+      X509ExtendedTrustManager delegateManager = null;
+      try {
+        final TrustManagerFactory tmf = TrustManagerFactory
+            .getInstance(TrustManagerFactory.getDefaultAlgorithm());
+        tmf.init(ks);
+        TrustManager[] tms = tmf.getTrustManagers();
+        // Iterate over the returned trust managers, looking for an instance of X509TrustManager.
+        // If found, use that as the delegate trust manager.
+        for (int i = 0; i < tms.length; i++) {
+          if (tms[i] instanceof X509ExtendedTrustManager) {
+            delegateManager = (X509ExtendedTrustManager) tms[i];
+            break;
+          }
+        }
+        if (delegateManager == null) {
+          throw new CertificateException(
+              "Instance delegateX509TrustManager is null. Failed to initialize");
+        }
+      } catch (Exception e) {
+        throw new CertificateException("Failed to initialize delegateX509TrustManager, error: "
+            + e.getMessage());
+      }
+      if (isClient) {
+        if (authType == VerificationAuthType.CertificateAndHostNameVerification
+            && (sslEngine == null || sslEngine.getSSLParameters() == null)) {
+          throw new CertificateException(
+              "SSLEngine or SSLParameters is null. Couldn't check host name");
+        }
+        if (sslEngine != null && sslEngine.getSSLParameters() != null) {
+          String algorithm = authType == VerificationAuthType.CertificateAndHostNameVerification
+              ? "HTTPS" : "";
+          SSLParameters sslParams = sslEngine.getSSLParameters();
+          sslParams.setEndpointIdentificationAlgorithm(algorithm);
```

Comment:
It is surprising this is being set, since these variables are normally set elsewhere are _read_ within the trust manager as configuration. This looks suspicious. It will also permanently change the parameters, which seems concerning. Have you verified that it is safe to disable endpoint identification here and it not impact any other code?

Reply:
Yes, thanks for the reminding! I have added some integration tests(`basicClientSideIntegrationTest` and `basicServerSideIntegrationTest`), and it looks like this `EndpointIdentificationAlgorithm` behaves as expected, without impacting other parts of the code. Please let me know if you find anything suspicious, thanks!

Reply:
"I tested it once" isn't what I really meant by verify. But since this on the engine, it is probably okay.